File: app/src/main/java/me/instabattle/app/managers/BattleManager.java
package me.instabattle.app.managers;

import android.location.Location;
import android.util.Log;

import java.util.List;

import me.instabattle.app.models.Battle;
import me.instabattle.app.settings.State;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

public class BattleManager {
    private static final String TAG = "BattleManager";
    private static final BattleService service = ServiceGenerator.createService(BattleService.class);

    public static void getAllAndDo(Callback<List<Battle>> callback) {
        Call<List<Battle>> call = service.getAll();
        call.enqueue(callback);
    }

    public static void getAndDo(Integer battleId, Callback<Battle> callback) {
        Call<Battle> call = service.get(battleId);
        call.enqueue(callback);
    }

    public static void getInRadiusAndDo(Location location, Double radius,
                                        Callback<List<Battle>> callback) {
        Call<List<Battle>> call = service.getInRadius(location.getLatitude(),
                location.getLongitude(), radius);
        call.enqueue(callback);
    }

    public static void createAndDo(Integer authorId, String name, Double latitude, Double longitude,
                                   String description, Integer radius, Callback<Battle> callback) {
        BattleService tokenService = ServiceGenerator.createService(BattleService.class, State.token);
        Log.d(TAG, "Sending token " + State.token);
        Call<Battle> call = tokenService.create(new Battle(authorId, name, description, latitude, longitude, radius));
        // FIXME photo actions
        call.enqueue(callback);
    }



    public static void updateAndDo(Integer battleId, String name,
                                   String description, Callback<Battle> callback) {
        Call<Battle> call = service.update(battleId, name, description);
        call.enqueue(callback);
    }

    interface BattleService {
        @GET("battles/{battle_id}")
        Call<Battle> get(@Path("battle_id") Integer battleId);

        @GET("battles/")
        Call<List<Battle>> getAll();

        @GET("battles/")
        Call<List<Battle>> getInRadius(@Query("latitude") Double lat,
                                       @Query("longitude") Double lon,
                                       @Query("radius") Double radius);

        // FIXME
        @PUT("battles/{battle_id}")
        Call<Battle> update(@Path("battle_id") Integer battleId,
                            @Body String name, @Body String description);

        @POST("battles/")
        Call<Battle> create(@Body Battle battle);

    }
}

File: app/src/main/kotlin/me/instabattle/app/managers/BattleManager.kt
package me.instabattle.app.managers

import android.location.Location

import me.instabattle.app.models.Battle
import me.instabattle.app.settings.State
import retrofit2.Call
import retrofit2.Callback
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

object BattleManager {
    private val service = ServiceGenerator.createService(BattleService::class.java)
    private var tokenService: BattleService? = null

    fun initTokenService() {
        tokenService = ServiceGenerator.createService(BattleService::class.java, State.token)
    }

    fun getAllAndDo(callback: Callback<List<Battle>>) {
        val call = service.all
        call.enqueue(callback)
    }

    fun getAndDo(battleId: Int?, callback: Callback<Battle>) {
        val call = service[battleId]
        call.enqueue(callback)
    }

    fun getInRadiusAndDo(location: Location, radius: Double?,
                         callback: Callback<List<Battle>>) {
        val call = service.getInRadius(location.latitude,
                location.longitude, radius)
        call.enqueue(callback)
    }

    fun createAndDo(authorId: Int?, name: String, latitude: Double?, longitude: Double?,
                    description: String, radius: Int?, callback: Callback<Battle>) {
        val call = tokenService!!.create(Battle(authorId, name, description, latitude, longitude, radius))
        call.enqueue(callback)
    }


    fun updateAndDo(battleId: Int?, name: String,
                    description: String, callback: Callback<Battle>) {
        val call = tokenService!!.update(battleId, name, description)
        call.enqueue(callback)
    }

    internal interface BattleService {

        @get:GET("battles/")
        val all: Call<List<Battle>>

        @GET("battles/{battle_id}")
        operator fun get(@Path("battle_id") battleId: Int?): Call<Battle>

        @GET("battles/")
        fun getInRadius(@Query("latitude") lat: Double?,
                        @Query("longitude") lon: Double?,
                        @Query("radius") radius: Double?): Call<List<Battle>>

        // FIXME
        @PUT("battles/{battle_id}")
        fun update(@Path("battle_id") battleId: Int?,
                   @Body name: String, @Body description: String): Call<Battle>

        @POST("battles/")
        fun create(@Body battle: Battle): Call<Battle>
    }
}
